File: solutions/_538_convert_bst_to_greater_tree/solution.js
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
var convertBST = function (root) {
  dfs(root, 0);
  return root;

  function dfs(node, sum) {
    if (node === null) return sum;
    node.val += dfs(node.right, sum);
    return dfs(node.left, node.val);
  }
};

module.exports = { convertBST };

/*
https://leetcode.com/problems/convert-bst-to-greater-tree/
*/

File: solutions/_538_convert_bst_to_greater_tree/solution.py
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def convertBST(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        self.dfs(root, 0)
        return root

    def dfs(self, node, total):
        if not node:
            return total
        node.val += self.dfs(node.right, total)
        return self.dfs(node.left, node.val)


# https://leetcode.com/problems/convert-bst-to-greater-tree/
